src/block.zig

```diff
@@ -0,0 +1,316 @@
+        acc[0] = .{
+            .name = "isa",
+            .type = ?*anyopaque,
+            .default_value = null,
+            .is_comptime = false,
+            .alignment = @alignOf(*anyopaque),
+        };
+        acc[1] = .{
+            .name = "flags",
+            .type = c_int,
+            .default_value = null,
+            .is_comptime = false,
+            .alignment = @alignOf(c_int),
+        };
+        acc[2] = .{
+            .name = "reserved",
+            .type = c_int,
+            .default_value = null,
+            .is_comptime = false,
+            .alignment = @alignOf(c_int),
+        };
+        acc[3] = .{
+            .name = "invoke",
+            .type = *const @Type(.{
+                .Fn = .{
+                    .calling_convention = .C,
+                    .alignment = @typeInfo(fn () callconv(.C) void).Fn.alignment,
+                    .is_generic = false,
+                    .is_var_args = false,
+                    .return_type = Return,
+                    .params = params,
+                },
+            }),
+            .default_value = null,
+            .is_comptime = false,
+            .alignment = @typeInfo(*const fn () callconv(.C) void).Pointer.alignment,
+        };
+        acc[4] = .{
+            .name = "descriptor",
+            .type = *Descriptor,
+            .default_value = null,
+            .is_comptime = false,
+            .alignment = @alignOf(*Descriptor),
+        };
+        std.debug.assert(!real_captures_info.is_tuple);
+        for (real_captures_info.fields, 0..) |capture, i| {
+            switch (capture.type) {
+                comptime_int => @compileError("capture should not be a comptime_int! try using @as"),
+                comptime_float => @compileError("capture should not be a comptime_float! try using @as"),
+                else => {},
+            }
+            acc[5 + i] = .{
+                .name = capture.name,
+                .type = capture.type,
+                .default_value = null,
+                .is_comptime = false,
+                .alignment = capture.alignment,
+            };
+        }
+        break :fields &acc;
+    };
+    return @Type(.{
+        .Struct = .{
+            .layout = .Extern,
+            .fields = fields,
+            .decls = &.{},
+            .is_tuple = false,
+        },
+    });
+}
+
+/// creates a block (on the heap, using malloc) of type T,
+// which should be the output of Block,
+// assigns the given captures
+// (which should be a struct literal with correct names)
+// and assigns the given function
+// (which should have the correct signature and C calling convention)
+// to the block.
+// Objective-C will free block pointers itself,
+// otherwise you can call free on it yourself.
+pub fn initBlock(comptime T: type, captures: anytype, blockFn: anytype) *T {
+    const allocator = std.heap.raw_c_allocator;
```

Comment:
Just want to confirm: I don't see a reason this has to be hardcoded to the c allocator, is there? Is there any reason we can't take an `Allocator` param?

Reply:
I suppose everything except the returned block could probably take an allocator. the reason it's on the heap at all is that when I passed a stack-allocated block to objective-c in the below snippet, I got an error from the objective-c runtime about `free` being called on memory that was not `malloc`'d.

https://github.com/ryleelyman/zig-cocoa-examples/blob/dce912b18bfd01a48992e40ecbb249b14c023e2b/src/examples/window.zig#L51-L81

Reply:
Ah interesting. I need to read up on exactly how blocks work, they must be taking ownership of some memory, which would make sense actually for closure data.